apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/InstanceHostManager.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright 2020 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Cloud.Spanner.Admin.Instance.V1;
+using Google.Cloud.Spanner.Common.V1;
+using Google.Protobuf.WellKnownTypes;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// Host manager for Spanner data client to get and cache the endpointUris for an instance.
+    /// </summary>
+    public sealed class InstanceHostManager
+    {
+        private ConcurrentDictionary<string, IList<string>> _instanceEndpointMapping =
+            new ConcurrentDictionary<string, IList<string>>();
+
+        /// <summary>
+        /// The default instance host manager.
+        /// </summary>
+        public static InstanceHostManager Default => new InstanceHostManager();
+
+        /// <summary>
+        /// Gets a host for the given instance.
+        /// If resource based routing is enabled (environment variable "GOOGLE_CLOUD_SPANNER_ENABLE_RESOURCE_BASED_ROUTING"),
+        /// this method will attempt to get the given instance endpointUris and return the first one.
+        /// The returning endpointUri will be cached to be used in future inquiries.
+        /// </summary>
+        /// <param name="projectId">The project ID. Must not be <c>null</c>.</param>
+        /// <param name="instanceId">The instance ID. Must not be <c>null</c>.</param>
+        /// <returns>The host for the given instance if found else null.</returns>
+        public async Task<string> GetHostAsync(string projectId, string instanceId)
+        {
+            if (SpannerConstants.IsResourceBasedRoutingEnabled)
+            {
+                if (!_instanceEndpointMapping.ContainsKey(instanceId))
+                {
+                    var instanceEndpointUris = await GetInstanceEndpointUrisAsync(projectId, instanceId).ConfigureAwait(false);
```

Comment:
This is still not thread safe. In between a call to `ContainsKey` in line 51 that returns false and line 54, some other thread may have come in asking for the same instance, `ContainsKey` would have also returned false, and now we have two threads asking the server for the same information and trying to store it on the same slot in the dictionary. And when I say two threads, I really mean many. It probably won't fail, but it beats the purpose of having a cache.
The advantage of having a `ConcurrentDictionary` is that it will take care of the thread safety for us, but we need to use the operations correctly. In this case we only need to use `GetOrAdd` operations, passing the key and a function that fetches the endpoints from the server. The dictionary will, atomically, see if it contains data for the key, and if it doesn't it will execute the function you provided and store the data obtained, and then either way, it will return the data associated to the key, again, all of that is done atomically by the dictionary. Note that the function is only execued if it needs to be.